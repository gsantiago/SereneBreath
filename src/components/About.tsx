import React from "react";
import { Popover } from "./Popover";

export function About({ isVisible }) {
  return (
    <Popover isVisible={isVisible} position="bottom">
      <section className="p-3">
        <p>made with love by @gsantiago</p>
      </section>
    </Popover>
  );
}
