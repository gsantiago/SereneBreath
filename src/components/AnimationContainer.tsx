import { animated, SpringValue, useSpringValue } from "@react-spring/web";

export const SIZE = 200;

export interface AnimationContainerProps {
  scale?: SpringValue<number>;
  children: React.ReactNode;
}

export function AnimationContainer({
  scale,
  children,
}: AnimationContainerProps) {
  const defaultScale = useSpringValue(1);

  return (
    <animated.div
      className="flex items-center justify-center "
      style={{ width: SIZE, height: SIZE, scale: scale ?? defaultScale }}
    >
      {children}
    </animated.div>
  );
}
