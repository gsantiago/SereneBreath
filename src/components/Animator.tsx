import React, { useEffect, useState } from "react";
import { AnimationProps } from "@/config/types";

export interface AnimatorProps
  extends Omit<AnimationProps, "state" | "isHolding"> {
  animation: React.FC<AnimationProps>;
}

export function Animator({
  animation: Animation,
  currentStep,
  pattern,
}: AnimatorProps) {
  const [state, setState] = useState<AnimationProps["state"]>("idle");

  const isHolding = currentStep === 1 || currentStep === 3;

  useEffect(() => {
    if (currentStep === 0 || currentStep === 1) {
      setState("inhaling");
    } else {
      setState("exhaling");
    }
  }, [currentStep]);

  return (
    <Animation
      currentStep={currentStep}
      pattern={pattern}
      state={state}
      isHolding={isHolding}
    />
  );
}
