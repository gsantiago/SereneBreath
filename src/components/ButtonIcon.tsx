import React from "react";

export interface ButtonIconProps
  extends Omit<React.HTMLProps<HTMLButtonElement>, "type"> {
  isActive?: boolean;
  position?: "top" | "bottom";
}

export function ButtonIcon({
  isActive,
  position = "top",
  ...props
}: ButtonIconProps) {
  const positionClass =
    position === "top" ? "rounded-t-md" : "rounded-b-md shadow-md";

  return (
    <button
      type="button"
      className={`px-2 pb-2 pt-2 text-3xl dark:text-white ${
        isActive
          ? `z-20 bg-white dark:bg-slate-800 ${positionClass}`
          : "rounded-md hover:bg-white hover:shadow-sm dark:hover:bg-slate-800"
      }`}
      {...props}
    />
  );
}
