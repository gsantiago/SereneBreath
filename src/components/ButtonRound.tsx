import React, { HTMLAttributes } from "react";

export function ButtonRound({
  style,
  ...props
}: HTMLAttributes<HTMLButtonElement>) {
  return (
    <button
      className="flex h-12 w-12 items-center justify-center rounded-full bg-teal-500 text-2xl text-white hover:bg-teal-600"
      {...props}
    />
  );
}
