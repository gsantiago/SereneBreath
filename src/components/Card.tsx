import { RingAnimation } from "@/animations/RingAnimation";
import { Animator } from "@/components/Animator";
import { Pattern } from "@/config/types";

export const WIDTH = 250;
const HEIGHT = "100%";

export interface CardProps {
  title: string;
  description: string;
  pattern: Pattern;
}

export function Card({ title, description, pattern }: CardProps) {
  return (
    <section
      className="overflow-hidden rounded-lg bg-white"
      style={{ width: WIDTH, height: HEIGHT, paddingBottom: 45 }}
    >
      <div className="flex h-1/2 w-full items-center justify-center bg-orange-100">
        <div style={{ transform: "scale(0.8)" }}>
          <Animator
            currentStep={3}
            pattern={[2, 2, 2, 2]}
            animation={RingAnimation}
          />
        </div>
      </div>
      <div className="p-5">
        <h1 className="text-xl font-bold sm:text-2xl">{title}</h1>
        <p className="my-1 text-xs text-gray-400">{pattern.join("-")}</p>
        <p className="text-sm text-gray-800 sm:text-base">{description}</p>
      </div>
    </section>
  );
}
