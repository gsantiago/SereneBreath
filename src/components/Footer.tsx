import { About } from "@/components/About";
import { ButtonIcon } from "@/components/ButtonIcon";
import { LanguageSelect } from "@/components/LanguageSelect";
import { useTranslation } from "@/hooks/useTranslation";

export interface FooterProps {
  isVisible: boolean;
  showAbout: boolean;
  onToggleAbout: () => void;
}

export function Footer({ isVisible, showAbout, onToggleAbout }: FooterProps) {
  const { t } = useTranslation();

  return (
    <footer
      className={`relative flex w-full justify-between text-xs transition-opacity dark:text-white ${
        isVisible ? "opacity-1" : "opacity-0"
      }`}
    >
      <div className="w-40">
        <LanguageSelect />
      </div>
      <ButtonIcon
        title={t("footer.about")}
        isActive={showAbout}
        position="bottom"
        onClick={onToggleAbout}
      >
        {infoIcon}
      </ButtonIcon>
      <About isVisible={showAbout} />
    </footer>
  );
}

const infoIcon = (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    fill="none"
    viewBox="0 0 24 24"
    strokeWidth={1.5}
    stroke="currentColor"
    className="h-6 w-6"
  >
    <path
      strokeLinecap="round"
      strokeLinejoin="round"
      d="M11.25 11.25l.041-.02a.75.75 0 011.063.852l-.708 2.836a.75.75 0 001.063.853l.041-.021M21 12a9 9 0 11-18 0 9 9 0 0118 0zm-9-3.75h.008v.008H12V8.25z"
    />
  </svg>
);
