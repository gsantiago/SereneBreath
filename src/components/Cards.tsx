import { Carousel } from "@/components/Carousel";
import { Card, WIDTH as CARD_WIDTH } from "@/components/Card";

import { useTranslation } from "@/hooks/useTranslation";
import { useStorage } from "@/hooks/useStorage";

import { techniques } from "@/config/techniques";
import { Technique } from "@/config/types";

export function Cards() {
  const { t } = useTranslation();

  const [technique, setTechnique] = useStorage("technique");
  const activeIndex = techniques.findIndex((t) => t.name === technique);

  return (
    <Carousel<Technique>
      activeIndex={activeIndex}
      itemWidth={CARD_WIDTH}
      items={[...techniques]}
      onChange={(item) => setTechnique(item.name)}
    >
      {({ item, isActive }) => (
        <Card
          title={t(`techniques.${item.name}.title`)}
          description={t(`techniques.${item.name}.description`)}
          pattern={item.pattern}
          isActive={isActive}
        />
      )}
    </Carousel>
  );
}
