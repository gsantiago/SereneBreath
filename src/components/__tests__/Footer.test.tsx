import { render, axe, screen, fireEvent } from "@/test-utils";
import { Footer, FooterProps } from "@/components/Footer";

const handlers = {
  onToggleAbout: vi.fn(),
};

const setup = (props?: Partial<FooterProps>) =>
  render(<Footer isVisible showAbout={false} {...handlers} {...props} />);

test("renders successfuly", () => {
  const { container } = setup();
  expect(container).toMatchSnapshot();
});

test("accessibility", async () => {
  const { container } = setup();
  expect(await axe(container)).toHaveNoViolations();
});

test("about button", () => {
  setup();

  const button = screen.getByTitle("About");
  fireEvent.click(button);

  expect(handlers.onToggleAbout).toHaveBeenCalled();
});
