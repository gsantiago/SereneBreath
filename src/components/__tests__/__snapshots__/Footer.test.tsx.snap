// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders successfuly 1`] = `
<div>
  <footer
    class="relative flex w-full justify-between text-xs transition-opacity dark:text-white opacity-1"
  >
    <div
      class="w-40"
    >
      <div>
        <label
          class="sr-only"
          for="locale"
        >
          Language
        </label>
        <select
          class="w-full cursor-pointer rounded-md border-gray-300 text-sm dark:text-gray-700"
          id="locale"
        >
          <option
            value="en"
          >
            English
          </option>
          <option
            value="pt"
          >
            PortuguÃªs
          </option>
        </select>
      </div>
    </div>
    <button
      class="px-2 pb-2 pt-2 text-3xl dark:text-white rounded-md hover:bg-white hover:shadow-sm dark:hover:bg-slate-800"
      title="About"
      type="button"
    >
      <svg
        class="h-6 w-6"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11.25 11.25l.041-.02a.75.75 0 011.063.852l-.708 2.836a.75.75 0 001.063.853l.041-.021M21 12a9 9 0 11-18 0 9 9 0 0118 0zm-9-3.75h.008v.008H12V8.25z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
    <div
      class="absolute  w-full rounded-md  bg-white py-2 shadow-md dark:bg-slate-800 sm:w-80 hidden right-0 bottom-full rounded-br-none z-10"
    >
      <section
        class="px-3 py-1"
      >
        <h1
          class="text-lg font-bold"
        >
          SereneBreath
        </h1>
        <p
          class="mt-2 text-xs"
        >
          SereneBreath.com is a small webapp to help you improve your breathing, relaxing and managing your stress.
        </p>
        <p
          class="mt-2 text-xs"
        >
          This project is open source and open for new contributions. Check out its repo for more info:
           
          <a
            class="text-blue-600 underline dark:text-sky-300"
            href="https://github.com/gsantiago/SereneBreath"
            target="_blank"
          >
            Github Repo.
          </a>
        </p>
        <p
          class="mt-2 text-xs"
        >
          English and bell audio files are from the great Breathly app, which also inspired this app.
           
          <a
            class="text-blue-600 underline dark:text-sky-300"
            href="https://breathly.app/"
            target="_blank"
          >
            Breathly Site.
          </a>
        </p>
        <div
          class="mt-5 text-center"
        >
          <small
            class="text-xs text-gray-400"
          >
            v0.0.1
          </small>
        </div>
      </section>
    </div>
  </footer>
</div>
`;
